### PR TITLE
Detect published versions for packages before publishing.

### PR DESCRIPTION
**Pull request by qinweiz**
Publishing should not just check the latest version, instead it should check all existing versions before publishing to avoid conflict.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/web-build-tools/146)

### CODE REVIEW

File: common/temp_modules/rush-rush-lib/package.json [hunk 1]
```diff
@@ -3,23 +3,24 @@
   "version": "0.0.0",
   "private": true,
   "dependencies": {
+    "@microsoft/node-library-build": ">=2.3.1 <3.0.0",
```
Comment (nickpape-msft) [inline common/temp_modules/rush-rush-lib/package.json]: This should be a rushDependency

File: common/temp_modules/rush-rush/package.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "private": true,
   "dependencies": {
+    "@microsoft/node-library-build": ">=2.3.1 <3.0.0",
```
Comment (nickpape-msft) [inline common/temp_modules/rush-rush/package.json]: same here

File: rush/rush-lib/src/utilities/Npm.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+export default class Npm {
+  public static publishedVersions(
+    packageName: string,
+    cwd: string,
+    env: { [key: string]: string }
+  ): string[] {
+    const versions: string[] = [];
+    try {
+      const packageTime: string = Utilities.executeCommandAndCaptureOutput('npm',
+        `view ${packageName} time --json`.split(' '),
```
Comment (nickpape-msft) [inline rush/rush-lib/src/utilities/Npm.ts]: I am considering the `package-json` package to collect this information for evergreen. Perhaps we should do the same thing here?

Comment (qinweiz) [inline rush/rush-lib/src/utilities/Npm.ts]: This is a utility class. You can make use of that class to collect information for package-json or whatever when you do it.

**Comment by nickpape-msft** [comment]
Reviewed 1 of 4 files at r1, 12 of 16 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/146)*

**Comment by nickpape-msft** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/146#-:-KgaMc3D-2oHfxwwHcuK:bnfp4nl)*

**Pull request merged**